build(scripts): allow a git ref for lsp_types

kylo252 opened the pull request:
- build(scripts): vim.loop was renamed to vim.uv
- build(scripts): allow a git ref for lsp types

USAGE
```sh
# this will overwrite runtime/lua/vim/lsp/types/protocol.lua
nvim -l scripts/lsp_types.lua gen

# specify a custom output file
nvim -l scripts/lsp_types.lua gen --build/new_lsp_types.lua

# specify a git reference from https://github.com/microsoft/lsprotocol
nvim -l scripts/lsp_types.lua gen --out runtime/lua/vim/lsp/types/protocol.lua --ref 2023.0.0a2 
```

kylo252 committed:
build(scripts): vim.loop was renamed to vim.uv

kylo252 committed:
build(scripts): use a tag reference for lsp types

justinmk commented on scripts/lsp_types.lua:
```diff
@@ -17,12 +17,12 @@ local function tofile(fname, text)
   vim.fn.system({
     'curl',
-    'https://raw.githubusercontent.com/microsoft/lsprotocol/main/generator/lsp.json',
+    'https://raw.githubusercontent.com/microsoft/lsprotocol/2023.0.0a2/generator/lsp.json',
```
the use-case of this script is "I want to update to latest LSP spec". with this change, that now requires editing this part of the script.

this could perhaps be a parameter that defaults to `main`.

kylo252 committed:
build(scripts): allow a git ref for lsp_types

USAGE:
nvim -l scripts/lsp_types.lua gen
nvim -l scripts/lsp_types.lua gen --build/new_lsp_types.lua
nvim -l scripts/lsp_types.lua gen --out runtime/lua/vim/lsp/types/protocol.lua --ref 2023.0.0a2

max397574 commented:
would it perhaps be worth writing a function somewhere that can be reused for arg parsing in scripts like this?
something like:
```lua
local function parse_args(flags)
  local arguments = {}
  for i = 1, #_G.arg do
    if current_arg:sub(1,3)=="--" then
      local arg_name = current_arg:sub(3,-1)
      if vim.tbl_contains(flags, arg_name) then
       arguments[arg_name]=true
      else
       arguments[arg_name] = _G.arg[i+i]
     end
    end
end
```

not at my main computer rn so I can't test this
also the variable names should be changed

kylo252 commented:
> would it perhaps be worth writing a function somewhere that can be reused for arg parsing in scripts like this? something like:

I partially based it on the parser function from `bump_deps.lua`
https://github.com/neovim/neovim/blob/e4da418ba8388e94bb186e3f9a2004ee1e96f1e5/scripts/bump_deps.lua#L416-L422

but kept the changes to minimum to stay within the scope of this PR

max397574 commented:
I was the original author of this script and also partially based my function of another script
but I think it doesn't make much sense if there are just slightly adjusted functions in every script
that's why I'm proposing to add a general one like the one I wrote above
I think this should cover most cases
if not let me know and I can extend it (perhaps also add an opts table to it?)

justinmk commented on scripts/lsp_types.lua:
```diff
@@ -184,17 +185,21 @@ function M.gen(opt)
-    index = index + 1
+local opt = {
+  output_file = 'runtime/lua/vim/lsp/types/protocol.lua',
+  ref = "main"
```
Is this script not covered by our linter?

```suggestion
  ref = 'main',
```

justinmk committed:
Update scripts/lsp_types.lua
